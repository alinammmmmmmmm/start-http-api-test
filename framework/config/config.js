const config = {
  url: 'https://bookstore.demoqa.com',
  credentials: {
    userName: 'samanta.maya',
    password: 'Administrator6666%!',
  },
};

export default config;
