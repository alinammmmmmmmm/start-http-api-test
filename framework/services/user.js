import supertest from 'supertest';
import config from '../../framework/config/config';

const { url } = config;

//контроллеры user
const user = {
  //функция создания юзера
  create_user: (payload) => {
    return supertest(url).post('/account/v1/user').send(payload);
  },
  //функция создания токена
  create_token: (payload) => {
    return supertest(url).post('/account/v1/generatetoken').send(payload);
  },

  //функция авторизации
  login: (payload) => {
    return supertest(url).post('/account/v1/authorized').send(payload);
  },
  //   получение userId
  async getUserId() {
    const payload = config.credentials;
    const res = await this.create_user(payload);
    return res.body.userID;
  },
  //получение токена
  async getToken() {
    const payload = config.credentials;
    const res = await this.create_token(payload);
    return res.body.token;
  },
  //функция получения инфы о юзере с передачей {UUID} и токена
  info: (token, userID) => {
    return supertest(url)
      .get(`/account/v1/user/${userID}`)
      .set('Authorization', `Bearer ${token}`)
      .send();
  },
  //функция удаления юзера
  delete: (token,userID) => {
    return supertest(url)
      .delete(`/account/v1/user/${userID}`)
      .set('Authorization', `Bearer ${token}`)
      .send();
  },
};
export default user;
