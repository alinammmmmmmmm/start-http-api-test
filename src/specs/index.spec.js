import config from '../../framework/config/config';
import { user, book } from '../../framework/services/user';
//есть библиотека преобразования в json-schema https://www.npmjs.com/package/joi-to-json-schema
//подключение библиотеки https://www.npmjs.com/package/jest-json-schema для валидации json-schema
import { matchers } from 'jest-json-schema';
expect.extend(matchers);

describe('POST /bookstore/v1/books', () => {
  test('Создание книги', async () => {
    const token = await user.tokenCache();
    const res = await book.create_book(
      { userId: config.userID, collectionOfIsbns: [{ isbn: config.isbn }] },
      token
    );
    expect(typeof token).toBe('string');
    expect(typeof res.body).toBe('object');
    expect(res.body).toMatchSchema(config.schema);
    expect(res.status).toEqual(201);
  });
});

describe('GET /bookstore/v1/book', () => {
  test('Получении информации о книге', async () => {
    const res = await book.info_book();
    expect(typeof res.body).toBe('object');
    expect(res.status).toEqual(200);

    //создание json снапшота ответа с помощью Jest,в аргументах можем проверить тип данных у динамических обьектов
    //обновление снапшота через npx jest --updateSnapshot имя_файла
    expect(res.body).toMatchSnapshot({
      pages: expect.any(Number),
    });
  });

  it.each([config.isbn, config.new_isbn])(
    `параметризированный тест c массивом данных`,
    async () => {
      const res = await book.info_book();
      expect(typeof res.body).toBe('object');
      expect(res.status).toEqual(200);
    }
  );
});

describe('PUT /bookstore/v1/books/{ISBN}', () => {
  test('Обновление книги', async () => {
    const res = await book.update_book({
      userId: config.userID,
      isbn: config.new_isbn,
    });
    expect(res.body.books[0].isbn).toBe('9781449325862');
    expect(res.status).toEqual(200);
  });
});

describe('DELETE /bookstore/v1/book', () => {
  test('Удаление книги parametric test', async () => {
    const res = await book.delete_book({
      isbn: config.new_isbn,
      userId: config.userID,
    });
    expect(res.status).toEqual(204);
  });
});
