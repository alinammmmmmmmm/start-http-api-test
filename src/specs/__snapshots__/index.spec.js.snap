// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`GET /bookstore/v1/book Получении информации о книге 1`] = `
Object {
  "author": "Richard E. Silverman",
  "description": "This pocket guide is the perfect on-the-job companion to Git, the distributed version control system. It provides a compact, readable introduction to Git for new users, as well as a reference to common commands and procedures for those of you with Git exp",
  "isbn": "9781449325862",
  "pages": Any<Number>,
  "publish_date": "2020-06-04T08:48:39.000Z",
  "publisher": "O'Reilly Media",
  "subTitle": "A Working Introduction",
  "title": "Git Pocket Guide",
  "website": "http://chimera.labs.oreilly.com/books/1230000000561/index.html",
}
`;
